src/connectivity/percolation.c

```diff
@@ -0,0 +1,281 @@
+    if (edge_count % 2 == 1) {
+        IGRAPH_ERROR("Invalid edge list, odd number of elements.", IGRAPH_EINVAL);
+    }
+    edge_count >>= 1;
+    IGRAPH_CHECK(igraph_vector_int_resize(output, edge_count));
+
+    for (igraph_integer_t i = 0; i < edge_count; i++) {
+        percolate_edge(&links, &sizes, &biggest, VECTOR(*edges)[2 * i], VECTOR(*edges)[2 * i + 1]);
+        VECTOR(*output)[i] = biggest;
+    }
+    igraph_vector_int_destroy(&sizes);
+    igraph_vector_int_destroy(&links);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_bond_percolation
+ * \brief calculates the bond-percolation curve of a graph.
+ * \param graph The input graph
+ * \param output output[i] is the size of the largest component after adding i+1 edges, will be created.
+ */
+
+igraph_error_t igraph_bond_percolation(
+        const igraph_t *graph,
+        igraph_vector_int_t * output,
+        igraph_vector_int_t * edge_indices) {
+
+    igraph_vector_int_t* internal_edge_indices = edge_indices;
+    igraph_vector_int_t new_edges;
+    igraph_bool_t generated_edges = false;
+
+    if (edge_indices == NULL) {
+        // Use random edge order
+        generated_edges = true;
+        igraph_integer_t size = igraph_ecount(graph);
+
+        IGRAPH_CHECK(igraph_vector_int_init_range(&new_edges, 0, size));
+
+        IGRAPH_CHECK(igraph_vector_int_shuffle(&new_edges));
+
+        IGRAPH_FINALLY(igraph_vector_int_destroy, &new_edges);
+        internal_edge_indices = &new_edges;
+    }
+
+    igraph_vector_int_t edgelist;
+    IGRAPH_CHECK(igraph_vector_int_init(&edgelist, 2 * igraph_vector_int_size(internal_edge_indices)));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &edgelist);
+
+    for (igraph_integer_t i = 0; i < igraph_vector_int_size(internal_edge_indices); i++) {
+        IGRAPH_CHECK(igraph_edge(graph,
+                                 VECTOR(*internal_edge_indices)[i],
+                                 &VECTOR(edgelist)[2 * i],
+                                 &VECTOR(edgelist)[2 * i + 1]));
+    }
+
+
+    IGRAPH_CHECK(edge_list_percolation(&edgelist, output));
+
+    igraph_vector_int_destroy(&edgelist);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (generated_edges) {
+        igraph_vector_int_destroy(&new_edges);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t percolate_site(const igraph_t *graph,
+                                     igraph_vector_int_t *links,
+                                     igraph_vector_int_t *sizes,
+                                     igraph_integer_t *biggest,
+                                     igraph_integer_t vertex) {
+
+    if (VECTOR(*sizes)[vertex] != 0) {
+        IGRAPH_ERROR("Vertex already added.", IGRAPH_EINVAL);
+    }
+
+    VECTOR(*sizes)[vertex] = 1;
+
+    igraph_vector_int_t neighbors;
+    IGRAPH_CHECK(igraph_vector_int_init(&neighbors, 0));
+    IGRAPH_FINALLY(igraph_vector_int_destroy, &neighbors);
```

Comment:
```suggestion
    IGRAPH_VECTOR_INT_INIT_FINALLY(igraph_vector_int_init(&neighbors, 0));
```

Please use the convenience macro whenever possible.